### PR TITLE
Kestrel doesn't like sync IO operations

### LINKED ISSUE #2164: DLNA: No folders or files listed when browsing
**Describe the bug**
When attempting to browse Jellyfin content (via VLC/DLNA), no folders or media items are listed.

**To Reproduce**

1. Open VLC
2. Click on `Universal Plug'n'Play`
3. Double click Jellyfin instance
4. No files or folders are displayed underneath Jellyfin

**Expected behavior**

List of libraries is displayed and can be expanded to view media items.

**Logs**

```
[14:59:43] [DBG] No matching device profile found. Connection=close, Accept=*/*, Host=192.168.1.141:8096
[14:59:47] [DBG] No matching device profile found. Content-Type=text/xml; charset="utf-8", Host=192.168.1.141:8096, User-Agent=6.2.9200 2/, UPnP/1.0, Portable SDK for UPnP devices/1.6.19, Content-Length=440, SOAPACTION="urn:schemas-upnp-org:service:ContentDirectory:1#Browse"
```

**Screenshots**

image

**System (please complete the following information):**
 - OS: Docker
 - Browser: DLNA - VLC
 - Jellyfin Version: 10.5-dev bc7cbfb21aced3ad72ddd98183ad457e50b48dc8
 - Installed Plugins: None
 - Reverse proxy: None

**Additional context**
Browsing and playing recorded media works fine on 10.4.3.
Related: #1181 #2132 #2134

### PR DESCRIPTION
**Pull request by Bond-009**
Ref: https://github.com/dotnet/aspnetcore/issues/7644

**Issues**
Should fix #2164

### CODE REVIEW

**Comment by PrplHaz4** [comment]
@Bond-009 Looks like I'm still seeing the same error with this PR:

```
P¶{ÝP>ªøoEwÎ@@4À¨À¨+ áZEè^t¤Pû(bHTTP/1.1 200 OK
Date: Sat, 18 Jan 2020 03:20:33 GMT
Content-Type: text/xml; charset=UTF-8
Server: Microsoft-NetCore/2.0, UPnP/1.0 DLNADOC/1.50
Content-Length: 591
Expires: 0
Access-Control-Allow-Headers: Accept, Accept-Language, Authorization, Cache-Control, Content-Disposition, Content-Encoding, Content-Language, Content-Length, Content-MD5, Content-Range, Content-Type, Date, Host, If-Match, If-Modified-Since, If-None-Match, If-Unmodified-Since, Origin, OriginToken, Pragma, Range, Slug, Transfer-Encoding, Want-Digest, X-MediaBrowser-Token, X-Emby-Authorization
Access-Control-Allow-Methods: GET, POST, PUT, DELETE, PATCH, OPTIONS
Access-Control-Allow-Origin: *
```
```xml
<?xml version="1.0" encoding="utf-8" standalone="yes"?>
<SOAP-ENV:Envelope SOAP-ENV:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/" xmlns:SOAP-ENV="http://schemas.xmlsoap.org/soap/envelope/">
<SOAP-ENV:Body>
<SOAP-ENV:Fault><faultcode>500</faultcode>
<faultstring>Synchronous operations are disallowed. Call ReadAsync or set AllowSynchronousIO to true instead.</faultstring>
<detail><UPnPError xmlns="urn:schemas-upnp-org:control-1-0"><errorCode>401</errorCode><errorDescription>Invalid Action</errorDescription></UPnPError></detail>
</SOAP-ENV:Fault></SOAP-ENV:Body></SOAP-ENV:Envelope>
```

**Comment by Bond-009** [comment]
There should be an exception in the logs, "Error processing control request" followed by a stacktrace, can you post that here?

**Comment by PrplHaz4** [comment]
I did not see any new log entries when testing this pr.

**Comment by PrplHaz4** [comment]
@Bond-009 Confirmed I can't get any additional logs or a stack trace with this PR.  From the linked issue - it sounds like we should be enabling snyc - not forcing async, no?

**Comment by Bond-009** [comment]
Ideally we should use the async methods, but if we want a quick and dirty solution we could enable sync operations in kestrel

**Comment by PrplHaz4** [comment]
I'm not sure this file is being hit as none of those debug statements are triggering logs.  The only log line I get is this.  I've been trying to add some debug statements but not getting anything useful...
```
[14:59:47] [DBG] No matching device profile found. Content-Type=text/xml; charset="utf-8", Host=192.168.1.141:8096, User-Agent=6.2.9200 2/, UPnP/1.0, Portable SDK for UPnP devices/1.6.19, Content-Length=440, SOAPACTION="urn:schemas-upnp-org:service:ContentDirectory:1#Browse"
```
which looks like it's coming from here, and seems to be erroneously returning no profile instead of the default profile (which might be why there's no other logs)....
https://github.com/jellyfin/jellyfin/blob/master/Emby.Dlna/DlnaManager.cs#L225

**Comment by PrplHaz4** [comment]
@Bond-009 It's late but I got it to work - it could use some cleanup, but wanted to leave it here so you didn't have to bang your head against the desk like I did...

**BaseControlHandler.cs**
```csharp
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Web;
using System.Text;
using System.Xml;
using Emby.Dlna.Didl;
using MediaBrowser.Controller.Configuration;
using MediaBrowser.Controller.Extensions;
using Microsoft.Extensions.Logging;

namespace Emby.Dlna.Service
{
    public abstract class BaseControlHandler
    {
        private const string NS_SOAPENV = "http://schemas.xmlsoap.org/soap/envelope/";

        protected readonly IServerConfigurationManager Config;
        protected readonly ILogger _logger;

        protected BaseControlHandler(IServerConfigurationManager config, ILogger logger)
        {
            Config = config;
            _logger = logger;
        }

        public ControlResponse ProcessControlRequest(ControlRequest request)
        {
            try
            {
                var enableDebugLogging = Config.GetDlnaConfiguration().EnableDebugLog;

                if (enableDebugLogging)
                {
                    LogRequest(request);
                }

                var response = ProcessControlRequestInternal(request);

                if (enableDebugLogging)
                {
                    LogResponse(response);
                }

                return response;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error processing control request");

                return new ControlErrorHandler().GetResponse(ex);
            }
        }

        private ControlResponse ProcessControlRequestInternal(ControlRequest request)
        {
            ControlRequestInfo requestInfo = null;

            using (var streamReader = new StreamReader(request.InputXml))
            {
                var readerSettings = new XmlReaderSettings()
                {
                    ValidationType = ValidationType.None,
                    CheckCharacters = false,
                    IgnoreProcessingInstructions = true,
                    IgnoreComments = true,
                    Async = true
                };

                using (var reader = XmlReader.Create(streamReader, readerSettings))
                {
                    requestInfo = ParseRequest(reader);
			_logger.LogInformation("RequestInfo: {0}", requestInfo);
                }
            }


		var ln = requestInfo.LocalName;
            _logger.LogDebug("Received control request {0}", ln);

            var result = GetResult(ln, requestInfo.Headers);

            var settings = new XmlWriterSettings
            {
                Encoding = Encoding.UTF8,
                CloseOutput = false,
                Async = true
            };

            StringWriter builder = new StringWriterWithEncoding(Encoding.UTF8);

            using (var writer = XmlWriter.Create(builder, settings))
            {
                writer.WriteStartDocumentAsync(true);

                writer.WriteStartElementAsync("SOAP-ENV", "Envelope", NS_SOAPENV);
                writer.WriteAttributeStringAsync(string.Empty, "encodingStyle", NS_SOAPENV, "http://schemas.xmlsoap.org/soap/encoding/");

                writer.WriteStartElementAsync("SOAP-ENV", "Body", NS_SOAPENV);
                writer.WriteStartElementAsync("u", ln + "Response", requestInfo.NamespaceURI);
                foreach (var i in result)
                {
                    writer.WriteStartElementAsync("", i.Key, requestInfo.NamespaceURI);
                    writer.WriteStringAsync(i.Value);
                    writer.WriteFullEndElementAsync();
                }
                writer.WriteFullEndElementAsync();
                writer.WriteFullEndElementAsync();

                writer.WriteFullEndElementAsync();
                writer.WriteEndDocumentAsync();
            }

            var xml = builder.ToString().Replace("xmlns:m=", "xmlns:u=");

            var controlResponse = new ControlResponse
            {
                Xml = xml,
                IsSuccessful = true
            };

            _logger.LogDebug(xml);

            controlResponse.Headers.Add("EXT", string.Empty);

            return controlResponse;
        }

        private ControlRequestInfo ParseRequest(XmlReader reader)
        {
            reader.MoveToContentAsync();
            reader.ReadAsync();
            // Loop through each element
            while (!reader.EOF && reader.ReadState == ReadState.Interactive)
            {
                if (reader.NodeType == XmlNodeType.Element)
                {
                    switch (reader.LocalName)
                    {
                        case "Body":
                            {
                                if (!reader.IsEmptyElement)
                                {
                                    using (var subReader = reader.ReadSubtree())
                                    {
                                        return ParseBodyTag(subReader);
                                    }
                                }
                                else
                                {
                                    reader.ReadAsync();
                                }
                                break;
                            }
                        default:
                            {
                                reader.SkipAsync();
                                break;
                            }
                    }
                }
                else
                {
                    reader.ReadAsync();
                }
            }

            return new ControlRequestInfo();
        }

        private ControlRequestInfo ParseBodyTag(XmlReader reader)
        {
            var result = new ControlRequestInfo();

            reader.MoveToContentAsync();
            reader.ReadAsync();

            // Loop through each element
            while (!reader.EOF && reader.ReadState == ReadState.Interactive)
            {
                if (reader.NodeType == XmlNodeType.Element)
                {
                    result.LocalName = reader.LocalName;
                    result.NamespaceURI = reader.NamespaceURI;

                    if (!reader.IsEmptyElement)
                    {
                        using (var subReader = reader.ReadSubtree())
                        {
                            ParseFirstBodyChild(subReader, result.Headers);
                            return result;
                        }
                    }
                    else
                    {
                        reader.ReadAsync();
                    }
                }
                else
                {
                    reader.ReadAsync();
                }
            }

            return result;
        }

        private void ParseFirstBodyChild(XmlReader reader, IDictionary<string, string> headers)
        {
            reader.MoveToContentAsync();
            reader.ReadAsync();

            // Loop through each element
            while (!reader.EOF && reader.ReadState == ReadState.Interactive)
            {
                if (reader.NodeType == XmlNodeType.Element)
                {
                    // TODO: Should we be doing this here, or should it be handled earlier when decoding the request?
                    headers[reader.LocalName.RemoveDiacritics()] = reader.ReadElementContentAsString();
                }
                else
                {
                    reader.ReadAsync();
                }
            }
        }

        private class ControlRequestInfo
        {
            public string LocalName;
            public string NamespaceURI;
            public IDictionary<string, string> Headers = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        }

        protected abstract IEnumerable<KeyValuePair<string, string>> GetResult(string methodName, IDictionary<string, string> methodParams);

        private void LogRequest(ControlRequest request)
        {
            if (!Config.GetDlnaConfiguration().EnableDebugLog)
            {
                return;
            }

            var originalHeaders = request.Headers;
            var headers = string.Join(", ", originalHeaders.Select(i => string.Format("{0}={1}", i.Key, i.Value)).ToArray());

            _logger.LogDebug("Control request. Headers: {0}", headers);
        }

        private void LogResponse(ControlResponse response)
        {
            if (!Config.GetDlnaConfiguration().EnableDebugLog)
            {
                return;
            }

            var originalHeaders = response.Headers;
            var headers = string.Join(", ", originalHeaders.Select(i => string.Format("{0}={1}", i.Key, i.Value)).ToArray());
            //builder.Append(response.Xml);

            _logger.LogDebug("Control response. Headers: {0}", headers);
        }
    }
}
```

**Comment by JustAMan** [comment]
Closing this in favour of https://github.com/jellyfin/jellyfin/pull/2316

**Comment by PrplHaz4** [comment]
@Bond-009 I tested https://github.com/jellyfin/jellyfin/pull/2297/commits/ac3b958c6791a48217cda5625a68ffeb18fbcae6 and it's working great for listing and playing movies.

File: Emby.Dlna/Service/BaseControlHandler.cs [hunk 1]
```diff
@@ -15,44 +16,34 @@ public abstract class BaseControlHandler
     {
         private const string NS_SOAPENV = "http://schemas.xmlsoap.org/soap/envelope/";
 
-        protected readonly IServerConfigurationManager Config;
-        protected readonly ILogger _logger;
+        protected IServerConfigurationManager Config { get; }
+        protected ILogger Logger { get; }
```
Comment (JustAMan) [inline Emby.Dlna/Service/BaseControlHandler.cs]: why?..

Comment (Bond-009) [reply Emby.Dlna/Service/BaseControlHandler.cs]: https://stackoverflow.com/questions/2149166/c-sharp-protected-property-or-field#2149178

File: Emby.Dlna/Service/BaseControlHandler.cs [hunk 2]
```diff
@@ -237,10 +228,7 @@ private void LogRequest(ControlRequest request)
                 return;
             }
 
-            var originalHeaders = request.Headers;
-            var headers = string.Join(", ", originalHeaders.Select(i => string.Format("{0}={1}", i.Key, i.Value)).ToArray());
-
-            _logger.LogDebug("Control request. Headers: {0}", headers);
+            Logger.LogDebug("Control request. Headers: {@Header}", request.Headers);
```
Comment (dkanada) [inline Emby.Dlna/Service/BaseControlHandler.cs]:
```suggestion
            Logger.LogDebug("Control request. Headers: {@Headers}", request.Headers);
```

**Pull request merged**